[Doc] generic listener update

PaulRMellor opened the pull request:
### Type of change

**Documentation**

All guides updated for the new generic listener config.

- Config examples and concepts updated throughout
- Updated introduction to listener configuration
- New appendix files for GenericKafkaListener:
    - GenericKafkaListener.adoc	 
    - GenericKafkaListenerConfiguration.adoc	 
    - GenericKafkaListenerConfigurationBootstrap.adoc	 
    - GenericKafkaListenerConfigurationBroker.adoc	 

_KafkaListeners schema and related schemas -- examples changed
We may want to remove the examples from those schemas?_

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

PaulRMellor commented:
For the old configuration, _KafkaListeners_ schema and related, I updated the examples (and thought I had reverted them).
But they have the new config. Maybe that's fine or maybe better to remove all the additional info from the old schemas completely to avoid confusion (my preference)

scholzj commented on documentation/api/io.strimzi.api.kafka.model.KafkaClusterSpec.adoc:
```diff
@@ -15,7 +15,10 @@ spec:
-      plain: {}
+      - name: plain
+        port: 9092
+        type: service
```
```suggestion
        type: internal
```

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc:
```diff
@@ -0,0 +1,217 @@
+By specifying a unique name and port for each listener,
+you can configure multiple listeners.
+
+.`type`
+
+The type is set as `internal`,
+or for external listeners, as `route`, `loadbalancer`, `nodeport` or `ingress`.
+
+internal:: You can configure internal listeners with or without encryption using the `tls` property.
+route:: Configures an external listener to expose Kafka using OpenShift `Routes` and the HAProxy router.
++
+A dedicated `Route` is created for every Kafka broker pod.
+An additional `Route` is created to serve as a Kafka bootstrap address.
+Kafka clients can use these `Routes` to connect to Kafka on port 443.
+ingress:: Configures an external listener to expose Kafka using Kubernetes `Ingress` and the {NginxIngressController}.
++
+A dedicated `Ingress` resource is created for every Kafka broker pod.
+An additional `Ingress` resource is created to serve as a Kafka bootstrap address.
+Kafka clients can use these `Ingress` resources to connect to Kafka on port 443.
++
+You must specify the hostnames used by the bootstrap and per-broker services
+using xref:type-GenericKafkaListenerConfigurationBootstrap-reference[`GenericKafkaListenerConfigurationBootstrap`]
+and xref:type-GenericKafkaListenerConfigurationBroker-reference[`GenericKafkaListenerConfigurationBroker`] properties.
++
+NOTE: External listeners using `Ingress` are currently only tested with the {NginxIngressController}.
+
+loadbalancer:: Configures an external listener to expose Kafka `Loadbalancer` type `Services`.
++
+A new loadbalancer service is created for every Kafka broker pod.
+An additional loadbalancer is created to serve as a Kafka _bootstrap_ address.
+Loadbalancers listen to connections on port 9094.
+nodeport:: Configures an external listener to expose Kafka using `NodePort` type `Services`.
++
+Kafka clients connect directly to the nodes of Kubernetes.
+An additional `NodePort` type of service is created to serve as a Kafka bootstrap address.
++
+When configuring the advertised addresses for the Kafka broker pods, Strimzi uses the address of the node on which the given pod is running.
+You can use `preferredNodePortAddressType` property to configure the xref:property-listener-config-preferredNodePortAddressType-{context}[first address type checked as the node address].
++
+NOTE: TLS hostname verification is not currently supported when exposing Kafka clusters using node ports.
```
I think it would be great to have here examples for each type. Especially with the Ingress one it would make it more obvious what is needed.

PaulRMellor replied:
Added. Also added source ranges + link for the loadbalancer config

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc:
```diff
@@ -0,0 +1,217 @@
+By specifying a unique name and port for each listener,
+you can configure multiple listeners.
+
+.`type`
```
Can we make all of these as links? They will be important to deeplink into.

PaulRMellor replied:
Changed all the _.`<>`_ titles in the schemas to _` ===_

scholzj commented on documentation/modules/con-securing-kafka-authentication.adoc:
```diff
@@ -96,6 +108,20 @@ use the `networkPolicyPeers` property.
 
 NOTE: Your configuration of Kubernetes must support ingress `NetworkPolicies` in order to use network policies in Strimzi.
+
+== Additional listener configuration options
+
+Add further listener configuration using `configuration` properties.
+You can use:
+
+* `bootstrap` and `brokers` properties to specify overrides for connecting to bootstrap and broker services
+* `brokerCertChainAndKey` to xref:proc-installing-certs-per-listener-str[configure a listener to use your own private key and server certificate]
+* `externalTrafficPolicy`, for `nodeport` and `loadbalancer` listeners, to specify whether a service routes external traffic to a local node or cluster-wide endpoints
+* `loadBalancerSourceRanges`, for `loadbalancer` listeners, to specify Classless Inter-Domain Routing (CIDR) ranges to connect to
+* `preferredNodePortAddressType`, for `nodeport` listeners, to specify the address type to use as the node address
+* `useServiceDnsDomain`, for `internal` listeners, to specify whether the Kubernetes service DNS domain is used
+
+For more information, see the xref:type-GenericKafkaListenerConfiguration-reference[GenericKafkaListenerConfiguration schema].
```
Does this really belong here? This section is about authentication. Wouldn't it be better to just say something like:

```
For more information about listener configuration and list of all options, see the xref:type-GenericKafkaListenerConfiguration-reference[GenericKafkaListenerConfiguration schema].
```

without listing all the options.

PaulRMellor replied:
It was added for anyone not so familiar with listener config that were other options available. 
But I guess a note with a single link is just as effective :+1:

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.KafkaListeners.adoc:
```diff
@@ -13,15 +13,23 @@ Each listener is defined when the `listeners` object has the given property.
-  external:
+  - name: plain
+    port: 9092
+    type: service
```
```suggestion
    type: internal
```

PaulRMellor replied:
Examples removed

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.KafkaListeners.adoc:
```diff
@@ -13,15 +13,23 @@ Each listener is defined when the `listeners` object has the given property.
+    tls: false
+  - name: plain
+    port: 9092
+    type: service
```
```suggestion
    type: internal
```

PaulRMellor replied:
Examples removed

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.KafkaListeners.adoc:
```diff
@@ -30,13 +38,9 @@ listeners:
-  plain: {}
+  - name: plain
+    port: 9092
+    type: service
```
```suggestion
    type: internal
```

PaulRMellor replied:
Examples removed

scholzj reviewed (commented):
Great work. Thanks.

I do not really like the updated docs for the old API options. I think that it is a bit weird since people who get there in the docs will be looking for the old options. I think we should do one of following:
* Just remove them without any replacement. Who needs them can find them in older version docs.
* Replace the text in the docs file with some deprecation notice and link to 0.19.0 documentation to make it easier to find it.
* Keep the old docs as they were -> they are still valid and work, just deprecated.

PaulRMellor commented:
> Great work. Thanks.
> 
> I do not really like the updated docs for the old API options. I think that it is a bit weird since people who get there in the docs will be looking for the old options. I think we should do one of following:
> 
> * Just remove them without any replacement. Who needs them can find them in older version docs.
> * Replace the text in the docs file with some deprecation notice and link to 0.19.0 documentation to make it easier to find it.
> * Keep the old docs as they were -> they are still valid and work, just deprecated.

Thanks Jakub. I think we can remove them then. I think it's less confusing to do that. I'll make sure each of the old schemas have a clear link to GenericKafkaListener. And maybe have that link to 0.19 too

PaulRMellor commented:
> Great work. Thanks.
> 
> I do not really like the updated docs for the old API options. I think that it is a bit weird since people who get there in the docs will be looking for the old options. I think we should do one of following:
> 
> * Just remove them without any replacement. Who needs them can find them in older version docs.
> * Replace the text in the docs file with some deprecation notice and link to 0.19.0 documentation to make it easier to find it.
> * Keep the old docs as they were -> they are still valid and work, just deprecated.

I removed all the old examples
I added a link to the 0.19 doc from KafkaListeners to se the examples.
I can add this link from the other deprecated schemas, but maybe too much.

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc:
```diff
@@ -0,0 +1,323 @@
+            - 88.208.76.87/32
+    #...
+----
+nodeport:: Configures an external listener to expose Kafka using `NodePort` type `Services`.
```
Do we need an empty line to separate it from the example?

scholzj commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListenerConfigurationBootstrap.adoc:
```diff
@@ -0,0 +1,196 @@
+  - name: external
+    port: 9094
+    type: nodeport
+    tls: false
```
We use TLS authentication below, so it might be best to set it to `true`.

```suggestion
    tls: true
```

scholzj reviewed (approved):
Some small nits. But LGTM otherwise.

ppatierno commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc:
```diff
@@ -0,0 +1,323 @@
+[id='property-generic-listeners-{context}']
+=== `listeners`
+You configure Kafka broker listeners using the `listeners` property in the `Kafka` resource.
+Each listener is defined as an array.
```
it's not true. The `listeners` field is an array of listener(s) and each listener is an object with some fields like `name`, `port`, and so on.

PaulRMellor replied:
I changed to:

> Listeners are defined as an array.

ppatierno commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc:
```diff
@@ -0,0 +1,323 @@
++
+A dedicated `Route` is created for every Kafka broker pod.
+An additional `Route` is created to serve as a Kafka bootstrap address.
+Kafka clients can use these `Routes` to connect to Kafka on port 443.
```
Should we highlight more this? I mean, the user could be confused by the fact that we mention port 443 but the example has 9094. So explaining that TLS on routes is enabled by default and it works on port 443 but the traffic is routed to the port 9094 configured in the example?

PaulRMellor replied:
Changed to:

> The client connects on port 443, the default router port, but traffic is then routed to the port you configure,
> which is `9094` in this example.

PaulRMellor replied:
Did the same for ingress:

> The client connects on port 443, the default controller port, but traffic is then routed to the port you configure,
> which is `9095` in this example.

ppatierno commented on documentation/api/io.strimzi.api.kafka.model.listener.arraylistener.GenericKafkaListener.adoc:
```diff
@@ -0,0 +1,323 @@
++
+A new loadbalancer service is created for every Kafka broker pod.
+An additional loadbalancer is created to serve as a Kafka _bootstrap_ address.
+Loadbalancers listen to connections on port 9094.
```
I was confused initially. It's not clear that you are referring to the example, are you? The sentence sounds like that loadbalancer always listens on port 9094 :-)

PaulRMellor replied:
It does.
I've changed to:

> Loadbalancers listen to the specified port number, which is port `9094` in the following example.

ppatierno commented on documentation/modules/proc-configuring-kafka-user.adoc:
```diff
@@ -16,7 +16,7 @@ For example:
 When you configure the `KafkaUser` authentication and authorization mechanisms, ensure they match the equivalent `Kafka` configuration:
 
-* `KafkaUser.spec.authentication` matches `Kafka.spec.kafka.listeners.*.authentication`
+* `KafkaUser.spec.authentication` matches `Kafka.spec.kafka.listeners.authentication`
```
the `authentication` field is not right under the `listeners` one because you can specify different authentication for every listener. I don't know in this case what's the best notation, maybe something like:

`Kafka.spec.kafka.listeners.<listener>.authentication`

@scholzj wdyt?

scholzj replied:
Well, I guess technically it would be like this: `Kafka.spec.kafka.listeners[*].authentication`, or?

ppatierno replied:
yeah, agree, using the JSON path notation is better.

PaulRMellor replied:
Thanks. Fixed in three places.

ppatierno commented on documentation/modules/proc-securing-kafka.adoc:
```diff
@@ -9,11 +9,11 @@ This procedure shows the steps involved in securing Kafka brokers when running S
 The security implemented for Kafka brokers must be compatible with the security implemented for the clients requiring access.
 
-* `Kafka.spec.kafka.listeners.*.authentication` matches `KafkaUser.spec.authentication`
+* `Kafka.spec.kafka.listeners.authentication` matches `KafkaUser.spec.authentication`
```
ditto as above

PaulRMellor replied:
Fixed
